system/jlib/jtrace.cpp

```diff
@@ -67,6 +67,117 @@ namespace opentel_trace = opentelemetry::trace;
+     * If the buffer size is not a multiple of 64 bits, the remaining bytes are filled with a partial random value.
+     *
+     * @param buffer A span of bytes to be filled with random data.
+     */
+    void GenerateRandomBuffer(opentelemetry::nostd::span<uint8_t> buffer) noexcept
+    {
+        auto bufferSize = buffer.size();
+        for (size_t i = 0; i < bufferSize; i += sizeof(uint64_t))
+        {
+            uint64_t value = generateRandom64();
+            if (i + sizeof(uint64_t) <= bufferSize)
+            {
+                memcpy(&buffer[i], &value, sizeof(uint64_t));
+            }
+            else
+            {
+                memcpy(&buffer[i], &value, bufferSize - i);
+            }
+        }
+    }
+
+    /**
+     * @brief Seeds the random number generator with a given seed sequence.
+     *
+     * This method initializes the internal state of the generator using the provided seed sequence.
+     *
+     * @param seedSequence A seed sequence used to initialize the generator's state.
+     */
+    void seed(std::seed_seq &seedSequence) noexcept
+    {
+        seedSequence.generate(reinterpret_cast<uint32_t *>(randomBase.data()),
+                            reinterpret_cast<uint32_t *>(randomBase.data() + randomBase.size()));
+    }
+
+    /**
+     * @brief Generates a random 64-bit integer.
+     *
+     * This method uses the xorshift128p algorithm to produce a random 64-bit integer.
+     * The algorithm is based on bitwise operations and is efficient for generating random numbers.
+     *
+     * @return A random 64-bit integer.
+     */
+    uint64_t generateRandom64() noexcept
+    {
+        // Uses the xorshift128p random number generation algorithm described in
+        // https://en.wikipedia.org/wiki/Xorshift
+        auto &state_a = randomBase[0];
+        auto &state_b = randomBase[1];
+        auto t        = state_a;
+        auto s        = state_b;
+        state_a       = s;
+        t ^= t << 23;        // a
+        t ^= t >> 17;        // b
+        t ^= s ^ (s >> 26);  // c
+        state_b = t;
+        return t + s;
+    }
+
+private:
+    std::array<uint64_t, 2> randomBase{};
+};
+
+class CustomIdGenerator : public opentelemetry::sdk::trace::IdGenerator
+{
+public:
+    CustomIdGenerator() : IdGenerator(true) {}
+
+    opentelemetry::trace::SpanId GenerateSpanId() noexcept override
+    {
+        uint8_t spanIdBuffer[trace_api::SpanId::kSize];
+        randomGenerator.GenerateRandomBuffer(spanIdBuffer);
+        return trace_api::SpanId(spanIdBuffer);
+    }
+    
+    opentelemetry::trace::TraceId GenerateTraceId() noexcept override
+    {
+        uint8_t traceIdBuffer[trace_api::TraceId::kSize];
+        randomGenerator.GenerateRandomBuffer(traceIdBuffer);
+        return trace_api::TraceId(traceIdBuffer);
+    }
+
+private:
+    inline static thread_local JTraceRandomGenerator randomGenerator;
```

Comment:
Minor concern whether this will be initialised on each thread start.  Reading the c++ standard I think it is only on 1st use.  I will verify though.

Reply:
I have confirmed it is only initialised on 1st use